File: Node.java
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;


public class Node {

    private String symbolIndex, command;
    private ArrayList<Edge> edges;
    private Pair<String, String> message;
    private Node previous;
    private int support, depth, inDegree, outDegree;
    private boolean isVisited;

    public Node(String symbolIndex, Pair<String, String> message, String command) {
        this.symbolIndex = symbolIndex; //Index of stored message
        edges = new ArrayList<Edge>();
        this.support = 0; //Node support
        this.message = message; //Message containing source and destination locations on chip
        this.command = command; //Read or write command stored
        // totalSupport = 0;
        this.depth = 0;
        this.isVisited = false;
        this.depth = 0;
        this.inDegree = 0; //Number of edges incoming
        this.outDegree = 0; //Number of edges outgoing
        this.previous = null;
    }

    //Deep copy constructor
    public Node(Node x) throws NoSuchMethodException, SecurityException, InstantiationException, IllegalAccessException,
            IllegalArgumentException, InvocationTargetException, ClassNotFoundException {

        this.symbolIndex = new String(x.symbolIndex);
        this.edges = new ArrayList<Edge>();

        for(Edge edge : x.getEdges()){
            this.edges.add(edge);
        }

        this.previous = x.previous;
        this.support = Integer.valueOf(x.support);
        this.command = new String(x.command);
        this.message = new Pair<>(x.message.getKey(), x.message.getValue());
        //this.totalSupport = x.totalSupport;
        this.depth = Integer.valueOf(x.depth);
        this.isVisited = Boolean.valueOf(x.isVisited);
        this.inDegree = Integer.valueOf(x.inDegree); 
        this.outDegree = Integer.valueOf(x.outDegree); 
    }

    public int getInDegree(){
        return inDegree;
    }

    public int getOutDegree(){
        return outDegree;
    }

    public void setInDegree(int x){
        inDegree = x;
    }

    public void setOutDegree(int x){
        outDegree = x;
    }

    public boolean isVisited(){ 
        return this.isVisited; 
    }

    public void setVisited(boolean isVisited){
        this.isVisited = isVisited;
    }
    
    public int getDepth(){
        return depth;
    }

    public void setDepth(int depth){
        this.depth = depth;
    }

    /*
    public int getTotalSupport(){
        return this.totalSupport;
    }

    public void setTotalSupport(int totalSupport){
        this.totalSupport = totalSupport;
    }
    */

    public int getSupport(){
        return this.support;
    }

    public void setSupport(int support){
        this.support = support;
    }

    public void setSymbolIndex(String symbolIndex){
        this.symbolIndex = symbolIndex;
    }

    public String getSymbolIndex(){
        return this.symbolIndex;
    }

    public Pair<String, String> getMessage(){
        return message;
    }

    public String getCommand(){
        return command;
    }

    public String toString(){
        return this.symbolIndex;
    }

    
    public Node getPrevious(){
        if(this.previous == null)
            return null;
            
        return this.previous;
    }

    public void setPrevious(Node previous){
        this.previous = previous;
    }

    public ArrayList<Edge> getEdges(){
        return this.edges;
    }

    public Edge getEdge(String id){
        for(Edge it : edges)
            if(it.getId().equals(id))
                return it;
        return null;
    }

    public Edge getEdge(Node destination){

        for(Edge edge : edges)
            if(edge.getDestination() == destination)
                return edge;
        
        return null;
    }

    //Edge is added, out degree is updated as well as the in degree of the destination node
    public void addEdge(Edge edge){
        edges.add(edge);
        this.outDegree = this.outDegree + 1;
        edge.getDestination().setInDegree(edge.getDestination().getInDegree()+1);
    }

    //Edge removal, update in degree of destination node and out degree of this node
    public void removeEdge(Edge edge){
        this.outDegree = this.outDegree - 1;
        edge.getDestination().setInDegree(edge.getDestination().getInDegree()-1);
        edges.remove(edge);
    }

    public void removeEdge(Node destination){
        for(Edge it: edges){
            if(it.getDestination() == destination){
                this.outDegree = this.outDegree - 1;
                it.getDestination().setInDegree(it.getDestination().getInDegree()-1);
                edges.remove(it);
                break;
            }
        }
    }

    public void clearEdges(){

        for(Edge it: edges){
            this.outDegree = this.outDegree - 1;
            it.getDestination().setInDegree(it.getDestination().getInDegree()-1);
        }

        edges.clear();
        
    }

    public int getEdgeSupport (Edge edge){
        return edges.get(edges.indexOf(edge)).getEdgeSupport();
    }

    public void setEdgeSupport(Edge edge, int support){
        edges.get(edges.indexOf(edge)).setEdgeSupport(support);
    }

    /*
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + Integer.valueOf(symbolIndex);
        return result;
    }
    */

    public boolean equals(Object obj) {

        if(this == obj)
            return true;
        if(obj == null)
            return false;
        if(getClass() != obj.getClass())
            return false;
        Node other = (Node) obj;
        if(!symbolIndex.equals(other.getSymbolIndex()))
            return false;
        
        return true;
    }

	public void removeDuplicateEdges() {

        int i = 0;
        while(i < edges.size()){

            Edge it = edges.get(0);
            int timesDetected = 0;
            Edge edge;
            int j = 0;

            while(j < edges.size()){
                edge = edges.get(j);
                if(edge == it){
                    if(timesDetected == 0){
                        timesDetected = timesDetected + 1;
                        continue;
                    }
                    else
                        edges.remove(j);
                }
                j++;
            }

            i++;
        }
	}

}

File: SeqMining-Python/src/graph/node.py
from multipledispatch import dispatch

from src.graph.edge import Edge
from src.logging import *
from z3 import *
# import pulp as pl

class Node:
    def __init__(self, graph, symbol_index, message, command, msg_type):
        self.symbol_index = symbol_index
        self.message = message
        self.command = command
        self.msg_type = msg_type
        self.edges = {}
        self.succ_nodes = []
        self.pred_nodes = []
        self.outgoing_edges = []
        self.incoming_edges = []
        self.support = 0
        self.depth = 0
        self.visited = False
        self.in_degree = 0
        self.out_degree = 0
        self.previous = None
        self.graph = graph
        self.z3var = Int(str(self.symbol_index))
        self.pulp_var = None

    def __str__(self):
        return self.symbol_index

    def get_symbol_index(self):
        return self.symbol_index

    def get_index(self):
        return self.symbol_index

    def get_message(self):
        return self.message
        
    def get_message_str(self):
        return '(' + self.get_source() + ', ' + self.get_destination() + ', ' + self.get_command() + ', ' + self.get_type() + ')'

    def get_type(self):
        return self.msg_type
    
    def get_source(self):
        return self.message[1]
    
    def get_destination(self):
        return self.message[2]

    def get_command(self):
        return self.command

    def get_edges(self):
        return self.edges

    def get_outgoing_edges(self):
        return self.outgoing_edges

    def get_incoming_edges(self):
        return self.incoming_edges

    def get_z3var(self):
        return self.z3var

    def get_succ_nodes(self):
        return self.succ_nodes
        # indices = []
        # for succ in self.succ_nodes:
        #     indices.append(succ.get_symbol_index())
        # return indices
    
    def get_pred_nodes(self):
        return self.pred_nodes

    def get_support(self):
        return self.support

    def get_depth(self):
        return self.depth

    def is_visited(self):
        return self.visited

    def get_in_degree(self):
        return self.in_degree

    def get_out_degree(self):
        return self.out_degree

    def get_previous(self):
        return self.previous

    def set_symbol_index(self, symbol_index):
        self.symbol_index = symbol_index

    def set_message(self, message):
        self.message = message

    def set_command(self, command):
        self.command = command

    def set_support(self, support):
        self.support = support

    def set_depth(self, depth):
        self.depth = depth

    def set_visited(self):
        self.visited = True

    def set_not_visited(self):
        self.visited = False

    def set_in_degree(self, in_degree):
        self.in_degree = in_degree

    def set_out_degree(self, out_degree):
        self.out_degree = out_degree

    def set_previous(self, previous):
        self.previous = previous

    def __str__(self):
        return self.symbol_index

    @dispatch(str)
    def get_edge(self, edge_id):
        return self.edges[edge_id]

    @dispatch(object)
    def get_edge(self, destination):
        return self.edges[self.symbol_index + '_' + str(destination)]

    def add_edge(self, edge):
        self.edges[str(edge)] = edge

    def add_succ(self, dest):
        self.succ_nodes.append(dest)

    def add_pred(self, src):
        self.pred_nodes.append(src)

    def add_incoming_edge(self, edge):
        self.pred_nodes.append(edge.get_source())
        self.incoming_edges.append(edge)

    def add_outgoing_edge(self, edge):
        self.succ_nodes.append(edge.get_destination())
        self.outgoing_edges.append(edge)

    def print_full(self):
        return "{:<7}".format(str(self.symbol_index) + ' : ') + '(' + self.get_source() + ', ' + self.get_destination()+', ' + self.get_command()+', '+self.get_type()+')'

    @dispatch(Edge)
    def remove_edge(self, edge):
        self.edges.pop(str(edge), None)

    @dispatch(object)
    def remove_edge(self, destination):
        self.edges.pop(self.symbol_index + '_' + str(destination), None)

    def clear_edges(self):
        self.edges.clear()

    def set_pulp_var(self, pulp_var):
        self.pulp_var = pulp_var

    def get_pulp_var(self):
        return self.pulp_var
